### PR TITLE
feat(opeadapt): add native chrome extension

### PR DESCRIPTION
**Pull request by KrishPatel13**
Progress:

The extension currently sends the documentInfo object: 
```
const documentInfo = {
    title: document.title,
    url: window.location.href,
    html: document.documentElement.outerHTML
};
```
On every change in the DOM. 

Example of how does the console receive messages:
image
so on

### CODE REVIEW

**Comment by abrichr** [comment]
Good work @KrishPatel13 !

Next up is connecting the extension to python and having the python module the updates.

**Comment by KrishPatel13** [comment]
@abrichr, Just to be clear, Currently I have a python module, browser.py that gets the updates in DOM and then it sends the reply_num (update count) and the whole documentInfo object via native messaging to the background.js script which logs in the console.

See here: 
https://github.com/MLDSAI/OpenAdapt/pull/230/files#diff-c51a7447dc4c4bc4b75e74075f2d00cdf140a026fe7dc02e6de6d64fcec06be5R44

My screenshot here is the exact description of the browser.py in action with background and content script :)
https://github.com/MLDSAI/OpenAdapt/pull/230#issue-1746148482

**Comment by KrishPatel13** [comment]
> Refer

From this, I interpret as, making a new python module that sends and then logs only the updated part of the DOM. 
Is that right ? :)

**Comment by abrichr** [comment]
https://learn.microsoft.com/en-us/windows/win32/ipc/pipes

**Comment by abrichr** [comment]
Super early but looks very relevant

**Comment by KrishPatel13** [comment]
@abrichr 

On researching and reading about pipes, I found that:

Creating a platform-independent solution that uses named pipes can be challenging, because named pipes work differently on Unix/Linux versus Windows systems. So essentially we would need different scripts for Windows and Unix/Linux and/or macOs.

A more platform-independent approach might be to use Python's built-in support for inter-process communication (IPC) via sockets. This approach works similarly to pipes but should function the same way across different operating systems. 

Or

Another way I can think of is to make this work by writing into a temp txt file. And then we want to run openadapt, we can get the data from it. If this works, we can measure the performance difference. If is huge, we can work on improving the performance. But, at least we would have something working the end :)


Please let me know your thoughts. 
THank you.

**Comment by abrichr** [comment]
Please remove the `.idea` directory 🙏

**Comment by KrishPatel13** [comment]
A Demo: https://mldsai.slack.com/archives/D056HTS5KEG/p1687015265698859

**Comment by KrishPatel13** [comment]
Next Steps:

- [ ] Integrate the named-piped scripts with browser extension to print the actual dom changes in extension
- [ ] Integrate the whole environment with openadapt.

File: native_chrome_extension/content.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * Send the document info to the background script
+*/
+function sendDocumentInfo() {
+    const documentInfo = extractDocumentInfo();
+    chrome.runtime.sendMessage({ type: 'document', documentInfo });
+}
+
+
+sendDocumentInfo();
+document.addEventListener('DOMSubtreeModified', sendDocumentInfo);
```
Comment (KrishPatel13) [inline native_chrome_extension/content.js:24]: Alternative to Mutation Observer but use DOM Subtree Modified.

Comment (abrichr) [reply native_chrome_extension/content.js:24]: Can you please list tradeoffs between these approaches?

File: native_chrome_extension/server.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        return 'quit'
+
+
+def server():
+    with NPopen('wt', name='test') as pipe:  # writable text pipe
+        print('Pipe Created = ', pipe.path)
+    
+        print('waiting for client to connect...')
+        stream = pipe.wait()
+        print('client connected')
```
Comment (abrichr) [inline native_chrome_extension/server.py:19]: Please use logger.info 🙏

**Comment by KrishPatel13** [comment]
Closing in favour of: https://github.com/OpenAdaptAI/OpenAdapt/pull/337

**Pull request closed without merging**